### PR TITLE
Implementation of poch and hyp1f1

### LINKED ISSUE #15670: feat (scipy.special): Add hyp1f1 confluent hypergeometric function
I would like to implement a logpdf for the sum of n=2 independent gamma distributions for pymc.
For the special case of n=2 Mathai's Method can be used like Eqn. 3 in [https://doi.org/10.1007/s00180-019-00924-9](https://doi.org/10.1007/s00180-019-00924-9) which involves the confluent hypergeometric function $\prescript{}{1}{\mathbf{F}}{_1}(a,b,z)$. 
- scipy.special ships with the hyp1f1 function.
- The C source code can be found [here](https://github.com/scipy/scipy/blob/main/scipy/special/cephes/hyperg.c).

What would be the way to go? Translate the code into pure Python/Numpy and let jax handle it with jit?
I think Tensorflow's XLA has no variant of the function, but one for the (~~confluent~~) hypergeometric function $\prescript{}{2}{\mathbf{F}}{_1}(a,b,z)$ [cf. here](https://www.tensorflow.org/probability/api_docs/python/tfp/math/hypergeometric/hyp2f1_small_argument)

Are there any plans (maybe also considering XLA developement) to add such a functionality?

### PR DESCRIPTION
**Pull request by renecotyfanboy**
I plan to implement some hypergeometric special functions for real parameter cases in JAX

- Pochhammer symbol, with handling of real + integer cases
- Confluent hypergeometric function 1F1 

Might close #15670

Some comments on the implementation : 
- I added custom jvp for both using known recursive relations 
- 1F1 is concordant with scipy's for values of b and z not too high and a not too low (if a <<1, b>~50 and z >~100, precision is lost, but these are extreme cases since 1F1 is an exponential function)
- Improvements can be found in [Pearson+2014](https://arxiv.org/abs/1407.7786), which I might add at some point 

I'm looking for feedback on the implementation of these two before digging in 2F1 and add recursions to improve 1F1 evaluation for large values of a and b

### CODE REVIEW

File: jax/_src/scipy/special.py [hunk 1]
```diff
@@ -1719,3 +1719,195 @@ def bernoulli(n: int) -> Array:
...
+  # Factorial definition when m is close to an integer, otherwise gamma definition.
+  z, m = promote_args_inexact("poch", z, m)
+
+  integer_branch = lambda x, n: lax.fori_loop(1,
+                                              lax.round(m).astype(jnp.int32),
+                                              lambda i, y: y * (z + i), z)
+
+  float_branch = lambda x, n: gamma(z + m) / gamma(z)
+
+  result = lax.cond(lax.abs(lax.round(m)-m) < 1e-10, integer_branch, float_branch, z, m)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1739]: Can you say more about the need for separate logic for integers? Having a `fori` loop in a `cond` branch will lead to far worse performance for non-scalar inputs, and I'm unclear on whether there is any benefit.

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -1719,3 +1719,195 @@ def bernoulli(n: int) -> Array:
...
+
+  return lax.select(m == 0., 1., result)
+
+
+@jnp.vectorize
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1744]: Is `jnp.vectorize` required here? I think the implementation below would be fine without it.

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -1719,3 +1719,195 @@ def bernoulli(n: int) -> Array:
...
+b ~>50 and z >~100 might be erroneous. Convention for a = b = 0 is 1, unlike
+in scipy's implementation.""")
+@jit
+def hyp1f1(a, b, x):
+  """
+  Implementation of the 1F1 hypergeometric function for real valued inputs
+  Backed by https://doi.org/10.48550/arXiv.1407.7786
+  There is room for improvement in the implementation using recursion to
+  evaluate lower values of hyp1f1 when a or b or both are > 60-80
+  """
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1900]: FYI this docstring is overwritten by `_wraps`. If it's only meant as a code comment, it's fine.

File: jax/_src/scipy/special.py [hunk 4]
```diff
@@ -1719,3 +1719,195 @@ def bernoulli(n: int) -> Array:
   k = jnp.arange(2, 50, dtype=bn.dtype)  # Choose 50 because 2 ** -50 < 1E-15
   q2 = jnp.sum(k[:, None] ** -m[None, :], axis=0)
   return bn.at[4::2].set(q1 * (1 + q2))
+
+
+@custom_derivatives.custom_jvp
+@jnp.vectorize
+@_wraps(osp_special.poch, module='scipy.special', lax_description="""\
+The JAX version only accepts positive and real inputs.""")
+@jit
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1728]: `jit(vectorize)` would be better than `vectorize(jit)` if you want to jit-compile the function by default.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for working on this! It all looks really good; a few inline comments below.

Also, please add these new APIs to the docs at https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipyspecial

File: jax/_src/scipy/special.py (hunk 1 above)
Comment (renecotyfanboy) [reply jax/_src/scipy/special.py:1739]: At first I thought it could result in better performance but since each branch are evaluated no matter what, it is indeed better to fall to the pure gamma definition, which works for integers

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!

**Comment by jakevdp** [comment]
It looks like there are errors when the new functions are run with x64 precision. You can reproduce them locally using an invocation like this
```
$ JAX_ENABLE_X64=1 JAX_NUM_GENERATED_CASES=90 pytest tests/lax_scipy_special_functions_test.py
```

File: jax/_src/scipy/special.py [hunk 5]
```diff
@@ -1719,3 +1719,180 @@ def bernoulli(n: int) -> Array:
...
+  Backed by https://doi.org/10.48550/arXiv.1407.7786
+  There is room for improvement in the implementation using recursion to
+  evaluate lower values of hyp1f1 when a or b or both are > 60-80
+  """
+  a, b, x = promote_args_inexact('hyp1f1', a, b, x)
+
+  result = lax.cond(lax.abs(x) < 100, _hyp1f1_serie, _hyp1f1_asymptotic, a, b, x)
+  index = (a == 0) * 1 + ((a == b) & (a != 0)) * 2 + ((b == 0) & (a != 0)) * 3
+
+  return lax.select_n(index, result, 1., jnp.exp(x), jnp.inf)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1891]: I think you need to replace `1.` here with `jnp.array(1, dtype=x.dtype)` to fix the type mismatch error.

Comment (renecotyfanboy) [reply jax/_src/scipy/special.py:1891]: It worked on my side!

**Comment by jakevdp** [comment]
It looks like that did the trick – last thing, could you squash your changes into a single commit? Thanks!

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
It looks like there are some NaNs in the gradient tests that are causing problems

**Comment by renecotyfanboy** [comment]
This looks like an edge case which is not handled, is there any way to get the corresponding values *xs locally ?

**Comment by jakevdp** [comment]
Sure, you should be able to reproduce the error locally with this:
```
JAX_NUM_GENERATED_CASES=25 pytest tests/lax_scipy_special_functions_test.py
```

**Comment by renecotyfanboy** [comment]
Okay thanks to you! I found out the reason of nans. It is basically numerical overflow of float32 due to a combination of entry parameters leading to high values of the gradient. These high values are cast to +/-jnp.inf and latter nans. This does not arise when enforcing double precision. How can I handle this? The easiest way would be to crop further the testing values to ensure this does not happen in tests

**Comment by jakevdp** [comment]
The input values are generated using the random generators specified in the `JAX_SPECIAL_FUNCTION_RECORDS` listing. For `hyp1f1`, it looks like you specified uniform values in the range `(0.5, 40)`.

**Comment by renecotyfanboy** [comment]
It should be good now, I restricted the parameter distribution to avoid overflow with x32 precision.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
All tests are passing – thanks for the contribution!
